dwave/optimization/symbols/_unaryop.pyx

```diff
@@ -0,0 +1,104 @@
+# cython: auto_pickle=False
+
+# Copyright 2025 D-Wave
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from cython.operator cimport typeid
+
+from dwave.optimization.libcpp.graph cimport ArrayNode
+from dwave.optimization.libcpp.nodes.unaryop cimport *
+from dwave.optimization._model cimport ArraySymbol, _register, _Graph
+
+# We need to be able to expose the available node types to the Python level,
+# so unfortunately we need this enum and giant if-branches
+cpdef enum _UnaryOpNodeType:
+    Absolute
+    Cos
+    Exp
+    Expit
+    Log
+    Logical
+    Negative
+    Not
+    Rint
+    Sin
+    Square
+    SquareRoot
+
+
+class _UnaryOpSymbol(ArraySymbol):
+    def __init_subclass__(cls, /, _UnaryOpNodeType node_type):
+        if node_type is _UnaryOpNodeType.Absolute:
+            _register(cls, typeid(AbsoluteNode))
+        elif node_type is _UnaryOpNodeType.Cos:
+            _register(cls, typeid(CosNode))
+        elif node_type is _UnaryOpNodeType.Exp:
+            _register(cls, typeid(ExpNode))
+        elif node_type is _UnaryOpNodeType.Expit:
+            _register(cls, typeid(ExpitNode))
+        elif node_type is _UnaryOpNodeType.Log:
+            _register(cls, typeid(LogNode))
+        elif node_type is _UnaryOpNodeType.Logical:
+            _register(cls, typeid(LogicalNode))
+        elif node_type is _UnaryOpNodeType.Negative:
+            _register(cls, typeid(NegativeNode))
+        elif node_type is _UnaryOpNodeType.Not:
+            _register(cls, typeid(NotNode))
+        elif node_type is _UnaryOpNodeType.Rint:
+            _register(cls, typeid(RintNode))
+        elif node_type is _UnaryOpNodeType.Sin:
+            _register(cls, typeid(SinNode))
+        elif node_type is _UnaryOpNodeType.Square:
+            _register(cls, typeid(SquareNode))
+        elif node_type is _UnaryOpNodeType.SquareRoot:
+            _register(cls, typeid(SquareRootNode))
+        else:
+            raise RuntimeError(f"unexpected _UnaryOpNodeType: {<object>node_type!r}")
+
+        cls._node_type = node_type
+
+    def __init__(self, ArraySymbol x):
+        cdef _Graph model = x.model
+        cdef _UnaryOpNodeType node_type = self._node_type
```

Comment:
We should change this so it raises a nicer error if the user tries to construct `_UnaryOpSymbol` directly